pkg/controllermanager/controller/bastion/bastion_control_test.go

```diff
@@ -0,0 +1,116 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bastion
+
+import (
+	"context"
+	"time"
+
+	operationsv1alpha1 "github.com/gardener/gardener/pkg/apis/operations/v1alpha1"
+	"github.com/gardener/gardener/pkg/logger"
+	mockclient "github.com/gardener/gardener/pkg/mock/controller-runtime/client"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+
+	"github.com/golang/mock/gomock"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+var _ = Describe("Controller", func() {
+	var (
+		mockCtrl   *gomock.Controller
+		mockClient *mockclient.MockClient
+		reconciler reconcile.Reconciler
+
+		bastionName = "bastion"
+		ctx         = context.TODO()
+		log         = logger.NewNopLogger()
+		maxLifetime = 12 * time.Hour
+	)
+
+	BeforeEach(func() {
+		mockCtrl = gomock.NewController(GinkgoT())
+		mockClient = mockclient.NewMockClient(mockCtrl)
+		reconciler = NewBastionReconciler(log, mockClient, maxLifetime)
+	})
+
+	Describe("Reconciler", func() {
+		It("should return nil because object not found", func() {
+			mockClient.EXPECT().Get(ctx, kutil.Key(bastionName), gomock.AssignableToTypeOf(&operationsv1alpha1.Bastion{})).Return(apierrors.NewNotFound(schema.GroupResource{}, ""))
+
+			result, err := reconciler.Reconcile(ctx, reconcile.Request{NamespacedName: types.NamespacedName{Name: bastionName}})
+			Expect(result).To(Equal(reconcile.Result{}))
+			Expect(err).To(Succeed())
+		})
+
+		It("should delete expired Bastions", func() {
+			mockClient.EXPECT().Get(ctx, kutil.Key(bastionName), gomock.AssignableToTypeOf(&operationsv1alpha1.Bastion{})).DoAndReturn(func(_ context.Context, _ client.ObjectKey, obj *operationsv1alpha1.Bastion) error {
+				created := time.Now().Add(-maxLifetime / 2)
+				expires := metav1.NewTime(time.Now().Add(5 * time.Second))
+				*obj = operationsv1alpha1.Bastion{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:              bastionName,
+						CreationTimestamp: metav1.NewTime(created),
+					},
+					Status: operationsv1alpha1.BastionStatus{
+						ExpirationTimestamp: &expires,
+					},
+				}
+
+				return nil
+			})
+
+			deleted := false
+			mockClient.EXPECT().Delete(ctx, gomock.AssignableToTypeOf(&operationsv1alpha1.Bastion{})).Do(func(_ context.Context, _ *operationsv1alpha1.Bastion) {
+				deleted = true
+			})
```

Comment:
I don't think this is the best way to assert that the `Delete` endpoint has been called exactly once. Please advise on how to do this better. ^^

Reply:
I guess this is being made sure by gomock out-of-the-box. There is https://pkg.go.dev/github.com/golang/mock/gomock#Call.Times which defaults to `1`, i.e., if there was another `Delete` call by the client the test will complain. Similarly, it will complain if there is no `Delete` call by the client. Hence, you can remove the `deleted` variable entirely.

Reply:
That is what I tried, but even with

```
mockClient.EXPECT().Delete(ctx, gomock.AssignableToTypeOf(&operationsv1alpha1.Bastion{})).Times(2)
```

the tests still succeed. Removing the EXPECT correcly fails the test with

> Unexpected call to *client.MockClient.Delete([context.TODO &Bastion{ObjectMeta:{bastion      0 2021-05-08 07:31:53.71348648 +0200 CEST m=-21599.974411513 <nil> <nil> map[] map[] [] []  []},Spec:BastionSpec{ShootRef:{},SeedName:nil,ProviderType:nil,SSHPublicKey:,Ingress:[]BastionIngressPolicy{},},Status:BastionStatus{Ingress:nil,Conditions:[]Condition{},LastHeartbeatTimestamp:<nil>,ExpirationTimestamp:2021-05-08 13:31:48.713486958 +0200 CEST m=-4.974411038,},}]) at /home/xrstf/gospace/src/github.com/gardener/gardener/pkg/controllermanager/controller/bastion/bastion_control.go:95 because: there are no expected calls of the method "Delete" for that receiver

Reply:
I forgot to call `mockCtrl.Finish()`.